### PR TITLE
FS-105 update intent agent

### PR DESCRIPTION
**Pull request by evpearce**
## Description
Simplify intent agent and make it use chat history as context

## Changelog
-  change intent prompt to be a small and simple system prompt with question and chat history being passed via the user prompt
- remove splitting questions up for things that would invole the datastore agent until we have a better idea of what we are going to use that for
- add a few examples in prompt foo tests 
- other good testing questions can be found under "Questions for Fact-Checking"  here https://scottlogic.atlassian.net/wiki/spaces/FS/pages/4424925193/Extracted+Example+Test+3

### CODE REVIEW

File: backend/promptfoo/intent_config.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+providers:
+  - id: openai:gpt-4o-mini
+    config:
+      temperature: 0
+
+prompts: file://promptfoo_test_runner.py:create_prompt
+
+tests:
+  - description: "questions directed towards the database look ups should have only 1 question -1"
```
Comment (gaganahluwalia) [inline backend/promptfoo/intent_config.yaml:11]: Should it be 'lookups'?

File: backend/promptfoo/intent_config.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      system_prompt_template: "intent-system"
+      user_prompt_template: "intent"
+      user_prompt_args:
+        chat_history: []
+        question: "Using Bloomberg.csv dataset give me the company with the best esg score"
+    assert:
+      - type: javascript
+        value: JSON.parse(output).questions.length === 0
+
+  - description: "verify that the correct company name is determinded from the chat history"
```
Comment (gaganahluwalia) [inline backend/promptfoo/intent_config.yaml:33]: Did you mean 'determined'?

File: backend/src/prompts/templates/intent-system.j2 [hunk 3]
```diff
@@ -0,0 +1,16 @@
+You are an expert in determinging the intent behind a user's question, breaking down complex questions into multiple simplier questions and forming stand alone questions based on context from chat history.
```
Comment (gaganahluwalia) [inline backend/src/prompts/templates/intent-system.j2:1]: A few typos : determining, simpler

File: backend/src/prompts/templates/intent-system.j2 [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+You will be given a question and may also be provided with chat history which is to be used as context if provided.
+
+
+- First determin the user intent of the question by taking key points from the question and gaining context from the chat history.
```
Comment (gaganahluwalia) [inline backend/src/prompts/templates/intent-system.j2:6]: A few typos : determine

File: backend/src/prompts/templates/intent-system.j2 [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+- Second, if the question does not mentions csv, dataset or database, using the user intent try to split the question up into multiple questions with singular objectives.
```
Comment (gaganahluwalia) [inline backend/src/prompts/templates/intent-system.j2:7]: mention instead of mention

File: backend/src/supervisors/supervisor.py [hunk 6]
```diff
@@ -16,16 +16,24 @@ async def solve_all(intent_json) -> None:
...
-    for question in questions:
         try:
+            question = intent_json["question"]
             (agent_name, answer, status) = await solve_task(question, get_scratchpad())
             update_scratchpad(agent_name, question, answer)
             if status == "error":
                 raise Exception(answer)
         except Exception as error:
             update_scratchpad(error=str(error))
+    else:
```
Comment (gaganahluwalia) [inline backend/src/supervisors/supervisor.py:27]: It makes sense to have a process_question func and then call it in both if and else

async def process_question(question):
        try:
            agent_name, answer, status = await solve_task(question, get_scratchpad())
            update_scratchpad(agent_name, question, answer)
            if status == "error":
                logger.info("status == error")
                raise Exception(answer)
        except Exception as error:
            update_scratchpad(error=str(error))
            

if not questions:
        question = intent_json.get("question")
        if question:
            await process_question(question)
    else:
        for question in questions:
            await process_question(question)

**Review by gaganahluwalia — COMMENTED** [review]
Please have a look at my suggestions.

File: backend/promptfoo/intent_config.yaml (hunk 1 above)
Comment (evpearce) [reply backend/promptfoo/intent_config.yaml:11]: Thanks - forgot that vs code doesn't have spell check, I'm so used to having it
I've run the files both through a spell checker now - thanks for pointing that out

File: backend/src/prompts/templates/intent-system.j2 (hunk 5 above)
Comment (IMladjenovic) [inline backend/src/prompts/templates/intent-system.j2:7]: Here it sort of feels like we are suggesting to the intent agent to split question up into multiple parts for the sake of it. We could possibly rephase this to give the IntentAgent the agency to decide whether or not it should split out into further parts.

File: backend/src/agents/intent_agent.py [hunk 7]
```diff
@@ -18,7 +18,6 @@
     tools=[],
 )
 class IntentAgent(Agent):
-
     async def invoke(self, utterance: str) -> str:
         user_prompt = engine.load_prompt("intent", question=utterance, chat_history=get_session_chat())
```
Comment (IMladjenovic) [inline backend/src/agents/intent_agent.py:22]:
```suggestion
        session_chat=get_session_chat()
        user_prompt = engine.load_prompt("intent", question=utterance, chat_history=session_chat if session_chat else "There is no chat history")

```
Would this improve the prompt?

File: backend/src/prompts/templates/intent-system.j2 (hunk 5 above)
Comment (evpearce) [reply backend/src/prompts/templates/intent-system.j2]: by "agency" to decide... what do you mean exactly
Just ask it to only split it up if it needs to I suppose
Do you have any suggestions?

Comment (IMladjenovic) [reply backend/src/prompts/templates/intent-system.j2]: I've experimented with some prompt engineering on your branch but can't come up with anything that I think is an improvement, we can resolve this thread

File: backend/src/supervisors/supervisor.py [hunk 8]
```diff
@@ -12,20 +12,25 @@
...
+        (agent_name, answer, status) = await solve_task(question, get_scratchpad())
+        update_scratchpad(agent_name, question, answer)
+        if status == "error":
+            raise Exception(answer)
+    except Exception as error:
+        update_scratchpad(error=str(error))
+
+
 async def solve_all(intent_json) -> None:
     questions = intent_json["questions"]
```
Comment (IMladjenovic) [inline backend/src/supervisors/supervisor.py:26]:
```suggestion
        questions = []
        if len(intent_json["questions"]) > 0:
            questions += intent_json["questions"]
        else:
            questions.append(intent_json["question"])
```
and then you just need to loop over questions. Might make this a bit more clear?

**Review by IMladjenovic — APPROVED** [review]

**Review by gaganahluwalia — APPROVED** [review]

**Pull request merged**